### ISSUE TITLE
Review all data structures to ensure that they maintain invariants while deserializing from serde

### ISSUE THREAD

**Issue by sanket1729** [rust-bitcoin/rust-bitcoin#1327]
Currently, it is possible to create `TaprootSpendInfo`, `TaprootBuilder`, and perhaps other data structures from serde::deserealize that break internal invariants. 

I think this basically boils down structure with at least one non-public field cannot use auto derives for `serde`

**Comment by sanket1729** [comment]
I think this is a P-high bug for serde users that accept these values from untrusted parties over the wire.

**Comment by sanket1729** [comment]
Will get to this tomorrow.

**Comment by tcharding** [comment]
Are you planning on doing this one still @sanket1729, it is marked for inclusion in this release. Is that what we want still?

**Comment by tcharding** [comment]
Oh I see you have done #1533 already, will that PR resolve this issue (its not marked as such)?

**Comment by Kixunil** [comment]
#1533 resolves only specific cases, we should still check other data structures.

**Comment by tcharding** [comment]
Would it be acceptable as a step forward to use do serde by through strings?

E.g.,
```rust
#[cfg(feature = "serde")]
crate::serde_utils::serde_string_impl!(Height, "a relative height value");
```

**Comment by Kixunil** [comment]
No, that'd add a huge bloat to binary formats.

**Comment by tcharding** [comment]
Cool, will do it properly then :)

**Comment by apoelstra** [comment]
Moving to 0.31. To actually close this I think we need to do something pretty big -- I talked with @tcharding on the phone and we're speculating about writing a tool which checks the entire set of public types against a master .toml file or something, which indicates whether these types should have serde, should have tostring/fromstr, should have consensus encoding, etc. And also checks for tests.

Maybe this is infeasible, but like, cargo-semver-checks does something awfully similar so we should take a look at what they're doing..

**Comment by Kixunil** [comment]
`cargo-semver-checks` is checking the json output of rustdoc which only contains public APIs, so it's impossible to check this using such approach.

IMO we should really just move all types with invariants into private submodules to make constructors private.

**Comment by apoelstra** [comment]
@Kixunil well, we can check a few things (though I agree, none of them are about internal invariants):

* Does everything that is "categorized" as being string-serializable have a ToString/FromStr implementation
* Ditto for consensus::Encodable/Decodable
* Ditto for serde deserialize/serialize

At the very least we could check that everything we intend to have these traits does, and that nothing has one but not the other, and that everything exposed by the public API is listed so we have a single source of truth for what every public type "should" have.

Then as for testing whether there are round-trip tests ... I suspect there's something clever we could do there, like having a separate `roundtrip_test` crate with exported round trip tests, and running `cargo doc` on that, or something.

>IMO we should really just move all types with invariants into private submodules to make constructors private.

We could also have a master list of types that do/do not have invariants, and using `cargo doc` we could ensure that these are/are not constructible outside of constructors. Maybe we could even enumerate the constructors and check them off an "invariants" checklist or something.

I'm just spitballing here ... but my conclusion is that the json output from rustdoc, which I hadn't ever considered using til now, could provide us with a ton of linting value.

**Comment by apoelstra** [comment]
I would also like to use it to draw a nice tree of our entire public API, along with crate boundaries etc ... I'll bet doing this exercise would've brought the sighash mess to our attention, for example.

**Comment by tcharding** [comment]
That image would be great of the just-now-proposed web site.

**Comment by sanket1729** [comment]
> IMO we should really just move all types with invariants into private submodules to make constructors private.

+1. And let the compiler do it's thing.

**Comment by tcharding** [comment]
> IMO we should really just move all types with invariants into private submodules to make constructors private.

I didn't understand this the first time I read it, now @sanket1729 has agreed with it then I'm definitely missing something. How does one use a public type if the constructor is private? I thought the invariant is supposed to be enforced by the constructor? Did you mean to describe some solution so no other code in the module can create an object without going through the constructor? Or am I being brain dead?

**Comment by sanket1729** [comment]
@tcharding I had to test it out just in case it failed, so here's a playground link.
 https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=9a1727caebae8121b5891446c96cca63 .


This will help that we don't have the struct constructor anywhere else in file. The causes for these bugs are mostly because other constructors like `FromStr`, `serde::DeSerialize`, `from_slice` and others directly construct the struct without the new constructor, thus skipping the invariant checks. This will help encapsulate the logic, and other users even in the same file will be forced to use the constructor thus mandating the struct invariants are respected.

**Comment by tcharding** [comment]
Ah yeah, thanks man. That is what I thought was meant but I was confused by the words :)

**Comment by Kixunil** [comment]
@apoelstra IMO the options you wrote are much more complicated and less reliable than just putting things into private submodules. But yes, we can get a ton of linting value that would be really nice to use for some other cases.

> How does one use a public type if the constructor is private? I thought the invariant is supposed to be enforced by the constructor?

I should've written "make *compiler-generated* constructors private".

**Comment by apoelstra** [comment]
I agree that putting stuff in private modules is a good idea. I'm just saying that we could use rustdoc to lint that we actually did this for all the types we were "supposed" to.

**Comment by tcharding** [comment]
How do we use rustdoc to prove that code doesn't build (that a type cannot be constructed)?

**Comment by apoelstra** [comment]
Rustdoc will tell us whether all the fields of a public `struct` are public, or whether an `enum` is public. One of these is true iff the type can be constructed without going through a constructor.
